Add docs on building your own service graphs (#3804)
grafana/tempo: docs/sources/tempo/metrics-generator/service_graphs/custom-service-graphs.md

---
title: Custom service graphs
menuTitle: Custom service graphs

> description: Build your own service graphs using the Node Graph visualization from Tempo generated metrics

Reviewer: Rewording slightly so we don't have a compound noun using the product name. 
```suggestion
description: Build your own service graphs using the Node Graph visualization from metrics generated by Tempo. 
```

weight: 300
---

# Custom service graphs

> Depending on your need, you might want to craft custom service graphs based on the metrics generated from metrics generator, so this page is exactly about that.  

Reviewer: ```suggestion
You can craft custom service graphs based on the metrics generated from metrics-generator.  
```

> ## Building custom graphs
> 
> 
> ### Prerequisites  

Reviewer: Great that you have a prereq section! Our other doc pages use "Before you begin," with the same kind of content. I've also changed the heading level up one and removed "Building custom graphs", since the content for building custom graphs really starts in the section section. 
```suggestion
## Before you begin  
```

> For custom service graphs we will use:
> - Tempo metrics generator
> - Grafana deployment of version v10 or higher 
> - Grafana [Node graph panel](https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/node-graph/)
> 
> Tempo metrics generator creates multiple metrics and one of which is `traces_service_graph_request_total` that holds:
> - relationships between services
> - total number of requests performed between services

Reviewer: ```suggestion
For custom service graphs, you'll use:
- Tempo metrics-generator
- Grafana deployment of version v10 or higher 
- Grafana [Node graph panel](https://grafana.com/docs/grafana/latest/panels-visualizations/visualizations/node-graph/)

The metrics-generator creates multiple metrics, including `traces_service_graph_request_total`. This metric holds:
- relationships between services
- total number of requests performed between services
```

> ### Creating Grafana dashboards
> Crate a new dashboard and add two variables:
> - data source of type `Prometheus`
> ![Custom service graph Grafana data source variable](../custom-service-graph-var-datasource.png)
> - service of type `Label values`, enable multi-value option
> ![Custom service graph service variable](../custom-service-graph-var-service.png)
> 
> 
> ### Add a panel
> 

Reviewer: ```suggestion
### Create one or more Grafana dashboards

1. In Grafana, [create a new dashboard](https://grafana.com/docs/grafana/latest/dashboards/build-dashboards/create-dashboard/).
1.  Add two variables:
  - data source of type `Prometheus`
  ![Custom service graph Grafana data source variable](../custom-service-graph-var-datasource.png)
  - service of type `Label values`, enable multi-value option
  ![Custom service graph service variable](../custom-service-graph-var-service.png)

### Add a panel

```

> 1. Create a panel with a single query called `edges`
> 2. Select your Prometheus data source with metrics from tempo metrics generator
> 3. Query using the following example:

Reviewer: ```suggestion
1. Create a panel with a single query called `edges`.
2. Select your Prometheus data source with metrics from the metrics-generator.
3. Query using the following example:
```

```
label_join(
  label_join(
    label_join(
      sum(increase(traces_service_graph_request_total{server=~"$service"}[5m])) by (server, client) > 0
      or
      sum(increase(traces_service_graph_request_total{client=~"$service"}[5m])) by (server, client) > 0,
    "source", "", "client"),
  "target", "", "server"), 
"id", "-", "server", "client")
```

> 4. Use instant query type
> 5. If you need to debug, switch to Table data visualization and read [Node graph panel](/docs/grafana/latest/panels-visualizations/visualizations/node-graph/) docs for more options and expected data shape.
> ![Custom service graph panel view](../custom-service-graph-with-query.png)

Reviewer: ```suggestion
1. Use Instant query type.
1. If you encounter issues, switch to the Table data visualization. Refer to the [Node graph panel](/docs/grafana/latest/panels-visualizations/visualizations/node-graph/) documentation for more options and expected data shape.

![Custom service graph panel view](../custom-service-graph-with-query.png)
```

> All data transformations are done with Prometheus `label_join` operators. Query explanation:
> - first `label_join` creates new field `id` which is required by Node graph panel
> - second and third `label_join` copy client and server labels to source and target respectively as this names are also required by Node graph panel
> - we query `traces_service_graph_request_total` twice with `OR` operator to get a combination of requests from and to all of selected services

Reviewer: ```suggestion
All data transformations are done with the Prometheus `label_join` operators. 

Explanation of the query: 
- The first `label_join` creates a new field `id` which is required by the Node graph panel.
- The second and third `label_join` copy the client and server labels to source and target respectively, as these names are also required by Node graph panel.
- `traces_service_graph_request_total` is queried twice with the `OR` operator to get a combination of requests from and to all of the selected services.
```

> This query does most of the job done, but unfortunately it leaves us with some limitations that can not be compensated even by Grafana transform data feature. Some limitations:
> - unable to add request stats to nodes and edges, such as req/sec and error rates
> - unable to add custom icons for nodes
> 
> This limitations can be overcome by wrapping up Prometheus query into a REST API. In which we have more flexibility with query results data transformations. You can use [Grafana Infinity data source](/grafana/plugins/yesoreyeram-infinity-datasource/) to visualize data from REST API.

Reviewer: ```suggestion
This query gets most of the job done. However, it's limitations can't be compensated for even using the Grafana transform data feature. 
These limitations include:
- unable to add request stats to nodes and edges, such as req/sec and error rates
- unable to add custom icons for nodes

These limitations can be overcome by wrapping the Prometheus query into a REST API, which provides more flexibility with query results data transformations.
You can use [Grafana Infinity data source](/grafana/plugins/yesoreyeram-infinity-datasource/) to visualize data from REST API.
```
